src/net/ossl.cc

```diff
@@ -802,8 +802,16 @@ class session : public enable_shared_from_this<session>, public session_impl {
             return make_ready_future<stop_iteration>(stop_iteration::no);
         case SSL_ERROR_SYSCALL:
         {
-            auto err = std::system_error(errno, std::system_category(), "System error encountered during SSL write");
-            return handle_output_error(err).then([] {
+            auto f = make_ready_future<>();
+            if (errno == 0) {
+                auto err = ossl_error::make_ossl_error("System error encountered during SSL write");
+                f = handle_output_error(err);
+            } else {
+                auto err = std::system_error(
+                    errno, std::system_category(), "System error encountered during SSL write");
+                f = handle_output_error(err);
+            }
+            return f.then([] {
```

Comment:
suggestion (non-blocking): not sure it's any cleaner, really, but maybe...

```suggestion
            [errno]() -> future<> {
                if (errno == 0) {
                    auto err = ossl_error::make_ossl_error("System error encountered during SSL write");
                    return handle_output_error(err);
                } else {
                    auto err = std::system_error(
                        errno, std::system_category(), "System error encountered during SSL write");
                    return handle_output_error(err);
                }
            }().then([] {
```